src/_purryOrderRules.ts

```diff
@@ -0,0 +1,175 @@
+ *
+ * There are 2 forms of the order rule, a simple one which only provides the
+ * projection function and assumes ordering is ascending, and a 2-tuple where
+ * the first element is the projection function and the second is the direction;
+ * this allows changing the direction without defining a more complex projection
+ * to simply negate the value (e.g. `(x) => -x`).
+ *
+ * We rely on the javascript implementation of `<` and `>` for comparison, which
+ * will attempt to transform both operands into a primitive comparable value via
+ * the built in `valueOf` function (and then `toString`). It's up to the caller
+ * to make sure that the projection is returning a value that makes sense for
+ * this logic.
+ *
+ * It's important to note that there is no built-in caching/memoization of
+ * projection function and therefore no guarantee that it would only be called
+ * once.
+ */
+export type OrderRule<T> =
+  | Projection<T>
+  | readonly [projection: Projection<T>, direction: keyof typeof COMPARATORS];
+
+type Projection<T> = (x: T) => Comparable;
+
+// We define the Comparable based on how JS coerces values into primitives when
+// used with the `<` and `>` operators.
+// @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#type_coercion
+type Comparable =
+  | ComparablePrimitive
+  | { [Symbol.toPrimitive](hint: string): ComparablePrimitive }
+  | { valueOf(): ComparablePrimitive }
+  | { toString(): string };
+
+//  Notice that `boolean` is special in that it is coerced as a number (0 for
+// `false`, 1 for `true`) implicitly.
+type ComparablePrimitive = number | string | boolean;
+
+/**
+ * Allows functions that want to handle a variadic number of order rules a
+ * a simplified API that hides most of the implementation details. The only
+ * thing users of this function need to do is provide a function that would take
+ * the data, and a compare function that can be used to determine the order
+ * between the items of the array.
+ * This functions takes care of the rest; it will parse rules, built the
+ * comparer, and manage the purrying of the input arguments.
+ */
+export function purryOrderRules<T>(
+  func: (data: ReadonlyArray<T>, compareFn: CompareFunction<T>) => unknown,
+  inputArgs: IArguments | ReadonlyArray<unknown>
+): unknown {
+  // We rely on casting blindly here, but we rely on casting blindly everywhere
+  // else when we call purry so it's fine...
+  const [dataOrRule, ...rules] = (
+    Array.isArray(inputArgs) ? inputArgs : Array.from(inputArgs)
+  ) as
+    | [data: ReadonlyArray<T>, ...rules: Readonly<NonEmptyArray<OrderRule<T>>>]
+    | Readonly<NonEmptyArray<OrderRule<T>>>;
+
+  if (!isOrderRule<T>(dataOrRule)) {
+    // dataFirst!
+
+    // @ts-expect-error [ts2556]: Typescript is failing to infer the type of rules
+    // correctly here after the type refinement above, rules should be non-empty
+    // when we get here.
+    const compareFn = orderRuleComparer(...rules);
```

Comment:
last time we shipped something with `ts-expect-error`, it broke someone's code because they were checking lib code and the comments were stripped. I'd prefer if we'd do `as any` instead ?

Reply:
That's what our typescript tests are for; they check that the typescript lib files compile without error, too. This code would be compiled to JS and only live as JS code in our distribution, where ts errors don't matter.

Reply:
just fyi, it was this issue:

- https://github.com/remeda/remeda/issues/163

not sure if it will come back ...

Reply:
aah, this was about using ts-expect-error on interfaces that result in declaration, which is different

- https://github.com/remeda/remeda/issues/163#issuecomment-1071215533